### PR TITLE
contrib/google.golang.org/grpc: add appsec monitoring of received rpc messages

### PR DESCRIPTION
**Pull request by Julio-Guerra**
First version of the appsec integration into gRPC by monitoring the
messages an RPC handler receives. To do so, the security rule address
`grpc.server.request.message` is introduced in the WAF, and a new
set of instrumentation gateway operations are now available in `grpcsec`.

More details in the specification at https://datadoghq.atlassian.net/wiki/spaces/APS/pages/2278064284/gRPC+Protocol+Support

### CODE REVIEW

File: internal/appsec/waf.go [hunk 1]
```diff
@@ -48,19 +51,33 @@ func registerWAF(rules []byte, timeout time.Duration) (unreg dyngo.UnregisterFun
...
 	// Register the WAF event listener
-	unregister := dyngo.Register(newWAFEventListener(waf, addresses, timeout))
+	var unregisterHTTP, unregisterGRPC dyngo.UnregisterFunc
+	if len(httpAddresses) > 0 {
+		log.Debug("appsec: registering http waf listening to addresses %v", httpAddresses)
+		unregisterHTTP = dyngo.Register(newWAFEventListener(waf, httpAddresses, timeout))
+	}
+	if len(grpcAddresses) > 0 {
+		log.Debug("appsec: registering grpc waf listening to addresses %v", grpcAddresses)
+		unregisterHTTP = dyngo.Register(newGRPCWAFEventListener(waf, grpcAddresses, timeout))
```
Comment (Julio-Guerra) [inline internal/appsec/waf.go:69]:
```suggestion
		unregisterGRPC = dyngo.Register(newGRPCWAFEventListener(waf, grpcAddresses, timeout))
```

File: internal/appsec/waf.go [hunk 2]
```diff
@@ -106,6 +123,53 @@ func newWAFEventListener(handle *waf.Handle, addresses []string, timeout time.Du
...
+			nbEvents uint32
+			logOnce  sync.Once
+		)
+		op.On(grpcsec.OnReceiveOperationFinish(func(_ grpcsec.ReceiveOperation, res grpcsec.ReceiveOperationRes) {
+			if atomic.LoadUint32(&nbEvents) == maxWAFEventsPerRequest {
+				logOnce.Do(func() {
+					log.Debug("appsec: ignoring the rpc message due to the maximum number of security events per grpc call reached")
+				})
+				return
+			}
```
Comment (Julio-Guerra) [inline internal/appsec/waf.go:143]: Move this out of this listener, in the handler operation listener.

Comment (Julio-Guerra) [reply internal/appsec/waf.go:143]: It is actually correct only there and invalid above.
cc @Hellzy

File: contrib/google.golang.org/grpc/appsec_test.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		t.Skip("appsec disabled")
+	}
+
+	rig, err := newRig(false)
+	require.NoError(t, err)
+	defer rig.Close()
+
+	client := rig.client
+
+	t.Run("unary rpc", func(t *testing.T) {
```
Comment (gbbr) [inline contrib/google.golang.org/grpc/appsec_test.go:32]:
```suggestion
	t.Run("unary", func(t *testing.T) {
```

File: contrib/google.golang.org/grpc/appsec_test.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+		finished := mt.FinishedSpans()
+		require.Len(t, finished, 1)
+
+		// The request should have the XSS attack attempt event (appsec rule id crs-941-100).
+		event := finished[0].Tag("_dd.appsec.json")
+		require.NotNil(t, event)
+		require.True(t, strings.Contains(event.(string), "crs-941-100"))
+	})
+
+	t.Run("streaming rpc", func(t *testing.T) {
```
Comment (gbbr) [inline contrib/google.golang.org/grpc/appsec_test.go:51]:
```suggestion
	t.Run("stream", func(t *testing.T) {
```

File: contrib/google.golang.org/grpc/server.go [hunk 5]
```diff
@@ -154,6 +157,10 @@ func UnaryServerInterceptor(opts ...Option) grpc.UnaryServerInterceptor {
 				}
 			}
 		}
+
```
Comment (gbbr) [inline contrib/google.golang.org/grpc/server.go:160]: nit: Can you please remove this empty line? See https://github.com/DataDog/dd-trace-go/wiki/Style-guidelines#general-guidelines point 3.

**Review by gbbr — COMMENTED** [review]
Nice work Julio. I just left some nits regarding styling for `contrib`.

File: contrib/google.golang.org/grpc/server.go (hunk 5 above)
Comment (Julio-Guerra) [reply contrib/google.golang.org/grpc/server.go]: Thanks. Done here and in the stream function too.

**Review by knusbaum — APPROVED** [review]
Looks great from tracing perspective. 
👍

**Review by Hellzy — APPROVED** [review]

File: internal/appsec/dyngo/instrumentation/grpcsec/tags.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+		span.SetTag("network.client.ip", ip)
+	}
+	for h, v := range httpsec.NormalizeHTTPHeaders(md) {
+		span.SetTag("grpc.metadata."+h, v)
+	}
+	return nil
+}
+
+// setEventSpanTags sets the security event span tags into the service entry span.
+func setEventSpanTags(span ddtrace.Span, events []json.RawMessage) error {
```
[the remark this answers is not in the document]
Comment (Julio-Guerra) [reply internal/appsec/dyngo/instrumentation/grpcsec/tags.go:48]: I agree - will do on the next iteration

**Pull request merged**